### PR TITLE
Use std::unique_ptr to manage memory in AlignedVector.

### PR DESCRIPTION
**Pull request by bangerth**
Relates to #12008. There is more to do here, but this is a good first step.

/rebuild

### CODE REVIEW

**Review by peterrum — APPROVED** [review]

**Review by tjhei — APPROVED** [review]
Looks good. Have you considered renaming data_begin to data? Maybe for a followup...

File: include/deal.II/base/aligned_vector.h [hunk 1]
```diff
@@ -719,7 +720,7 @@ inline AlignedVector<T>::AlignedVector()
 
 template <class T>
 inline AlignedVector<T>::AlignedVector(const size_type size, const T &init)
-  : data_begin(nullptr)
+  : data_begin(nullptr, [](T *) { Assert(false, ExcInternalError()); })
```
Comment (drwells) [inline include/deal.II/base/aligned_vector.h:723]: This doesn't seem right. Suppose someone calls this with `size = 0` and never resizes the vector. The object will then be in a state where its destructor unconditionally aborts.

A better solution would be to just have an empty function body for the deleter - it doesn't need to do anything.

**Review by drwells — CHANGES_REQUESTED** [review]
I have a comment on the deleters.

File: include/deal.II/base/aligned_vector.h (hunk 1 above)
Comment (masterleinad) [reply include/deal.II/base/aligned_vector.h:723]: ... also destructors should not throw (which might happen if `disable_abort_on_exception` is enabled).

Comment (drwells) [reply include/deal.II/base/aligned_vector.h:723]: I should have checked the standard before writing that - it looks like unique_ptr doesn't call the deleter when it has a `nullptr`. The following code works:
```cpp
#include <iostream>
#include <memory>

template <typename T>
struct AlignedVector
{
  std::unique_ptr<T[], void (*)(T *)> data_begin;

    AlignedVector()
        : data_begin(nullptr, [](T *) { std::cout << "BBB\n"; std::abort();})
    {
        std::cout << "AAA\n";
    }
};

int main()
{
    AlignedVector<int> vec;
}
```

**Comment by bangerth** [comment]
The tester seems stuck. Let me rebase...

/rebuild

**Comment by bangerth** [comment]
I've also added a test to check the cases @drwells was concerned about.

**Comment by bangerth** [comment]
On 4/5/21 10:28 AM, Timo Heister wrote:
> Looks good. Have you considered renaming data_begin to data? Maybe for a 
> followup...

Yes, actually that patch is already ready and waiting for this one to be
merged :-) The asymmetry between `data_begin` (pointing to memory it owns) and 
`data_end` (not owning anything) bothered me.

**Pull request merged**